cmrf_views/src/Plugin/views/query/API4.php

```diff
@@ -0,0 +1,525 @@
+   */
+  protected $core;
+
+  /**
+   * @var \Drupal\views\ViewsData
+   */
+  protected $viewsData;
+
+  /**
+   * API constructor.
+   *
+   * @param array                   $configuration
+   * @param                         $plugin_id
+   * @param                         $plugin_definition
+   * @param \Drupal\cmrf_core\Core  $core
+   * @param \Drupal\views\ViewsData $views_data
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    Core $core,
+    ViewsData $views_data
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->core      = $core;
+    $this->viewsData = $views_data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(
+    ContainerInterface $container,
+    array $configuration,
+    $plugin_id,
+    $plugin_definition
+  ) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('cmrf_core.core'),
+      $container->get('views.views_data')
+    );
+  }
+
+  /**
+   * Method to trick views, because it expects an SQL backend.
+   *
+   * @param      $table
+   * @param null $relationship
+   *
+   * @return string
+   */
+  public function ensureTable($table, $relationship = NULL) {
+    return $table;
+  }
+
+  /**
+   * Method to trick views, because it expects an SQL backend.
+   *
+   * @param        $table
+   * @param        $field
+   * @param string $alias
+   * @param array  $params
+   *
+   * @return mixed
+   */
+  public function addField($table, $field, $alias = '', $params = []) {
+    // We check for this specifically because it gets a special alias.
+    if ($table == $this->view->storage->get('base_table') && $field == $this->view->storage->get('base_field') && empty($alias)) {
+      $alias = $this->view->storage->get('base_field');
+    }
+
+    if (!$alias && $table) {
+      $alias = $table . '_' . $field;
+    }
+
+    // Make sure an alias is assigned
+    $alias = $alias ? $alias : $field;
+
+    // Create a field info array.
+    $field_info = [
+        'field' => $field,
+        'table' => $table,
+        'alias' => $alias,
+      ] + $params;
+
+    // Test to see if the field is actually the same or not. Due to
+    // differing parameters changing the aggregation function, we need
+    // to do some automatic alias collision detection:
+    $base = $alias;
+    $counter = 0;
+    while (!empty($this->fields[$alias]) && $this->fields[$alias] != $field_info) {
+      $field_info['alias'] = $alias = $base . '_' . ++$counter;
+    }
+
+    if (empty($this->fields[$alias])) {
+      $this->fields[$alias] = $field_info;
+    }
+
+    // Keep track of all aliases used.
+    $this->fieldAliases[$table][$field] = $alias;
+
+    return $alias;
+  }
+
+  /**
+   * Returns the alias for the given field added to $table.
+   *
+   * @access protected
+   *
+   * @see \Drupal\cmrf_views\Plugin\views\query\API::addField()
```

Comment:
I guess this should reference `API4` instead of `API` so it might be replaced by `@see addField()`

Reply:
It should, I'll change that.